ql/src/java/org/apache/hadoop/hive/ql/parse/type/HiveFunctionHelper.java

```diff
@@ -0,0 +1,549 @@
+              TypeConverter.convert(inputs.get(2).getType()),
+              TypeConverter.convert(inputs.get(3).getType())));
+    } else if (genericUDF instanceof GenericUDFIn) {
+      // We're only considering the first element of the IN list for the type
+      assert inputs.size() > 1;
+      targetType = FunctionRegistry.getCommonClassForComparison(
+          TypeConverter.convert(inputs.get(0).getType()),
+          TypeConverter.convert(inputs.get(1).getType()));
+    }
+
+    if (targetType != null) {
+      List<RexNode> newInputs = new ArrayList<>();
+      // Convert inputs if needed
+      for (int i = 0; i < inputs.size(); ++i) {
+        RexNode input = inputs.get(i);
+        TypeInfo inputTypeInfo = TypeConverter.convert(input.getType());
+        RexNode tmpExprNode = input;
+        if (TypeInfoUtils.isConversionRequiredForComparison(targetType, inputTypeInfo)) {
+          if (isIN || isCompare) {
+            // For IN and compare, we will convert requisite children
+            tmpExprNode = convert(targetType, input);
+          } else if (isBetween) {
+            // For BETWEEN skip the first child (the revert boolean)
+            if (i > 0) {
+              tmpExprNode = convert(targetType, input);
+            }
+          } else if (isNumeric) {
+            // For numeric, we'll do minimum necessary cast - if we cast to the type
+            // of expression, bad things will happen.
+            PrimitiveTypeInfo minArgType = ExprNodeDescUtils.deriveMinArgumentCast(inputTypeInfo, targetType);
+            tmpExprNode = convert(minArgType, input);
+          } else {
+            throw new AssertionError("Unexpected " + targetType + " - not a numeric op or compare");
+          }
+        }
+
+        newInputs.add(tmpExprNode);
+      }
+      return newInputs;
+    }
+    return inputs;
+  }
+
+  private RexNode convert(TypeInfo targetType, RexNode input) throws SemanticException {
+    if (targetType.getCategory() == Category.PRIMITIVE) {
+      return RexNodeTypeCheck.getExprNodeDefaultExprProcessor(rexBuilder)
+          .createConversionCast(input, (PrimitiveTypeInfo) targetType);
+    } else {
+      StructTypeInfo structTypeInfo = (StructTypeInfo) targetType; // struct
+      RexCall call = (RexCall) input;
+      List<RexNode> exprNodes = new ArrayList<>();
+      for (int j = 0; j < structTypeInfo.getAllStructFieldTypeInfos().size(); j++) {
+        exprNodes.add(
+            convert(
+                structTypeInfo.getAllStructFieldTypeInfos().get(j), call.getOperands().get(j)));
+      }
+      return rexBuilder.makeCall(SqlStdOperatorTable.ROW, exprNodes);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public RexNode getExpression(String functionText, FunctionInfo fi,
+      List<RexNode> inputs, RelDataType returnType)
+      throws SemanticException {
+    // See if this is an explicit cast.
+    RexNode expr = RexNodeConverter.handleExplicitCast(
```

Comment:
I think it would be probably cleaner to create a RexNodeConverter; and pass a rexBuilder to it
after that :  remove the static from those methods - and the rexBuilder passed to every invocation of it...

Reply:
Tbh this was in my TODO list: many of those methods are not needed anymore in `RexConverter`. Hence my idea is to simply move them into `HiveFunctionHelper`. However, I did not want to do it as part of that patch to avoid further refactoring, as it was going to be shown as a bunch of removed code plus new code. I have created https://issues.apache.org/jira/browse/HIVE-23227 to tackle this.